app/components/ParagraphList.tsx

```diff
@@ -9,92 +8,229 @@ import { HIGHLIGHT_COLORS } from "~/resources/constants";
+              />
+            ))}
+        </div>
+      </HighlightContext.Provider>
+    </PrinciplesContext.Provider>
+  );
+}
+
+const MemoedParagraph = memo(function Paragraph({
+  paragraph,
+}: Readonly<{ paragraph: Paragraph }>) {
+  const principlesToShow = useContext(PrinciplesContext);
+
+  const groupedAbsaetze = useMemo(
+    () =>
+      groupAbsaetzeWithoutRelevantPrinciples(
+        paragraph.Absaetze,
+        principlesToShow,
+      ),
+    [paragraph.Absaetze, principlesToShow],
+  );
+
+  // If there are no Absätze with relevant principles, don't show the paragraph
+  if (groupedAbsaetze.every((absatzGroup) => absatzGroup instanceof Array))
     return null;
-  }
-  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-  const text = children.props.children as string;
-  return text.split(/(\[\d])/g);
-};
 
-const RemoveHighlight = ({ children }: { children: ReactNode }) => {
-  const parts = extractTextParts(children);
-  if (!parts) return null;
-  return parts[1] ? parts[0] : parts.join("");
+  const getAbsatzGroupTitle = (absatzGroup: AbsatzWithNumber[]) =>
+    absatzGroup.length > 1
+      ? `(${absatzGroup[0].number}) – (${absatzGroup[absatzGroup.length - 1].number})`
+      : `(${absatzGroup[0].number})`;
+
+  return (
+    <div key={paragraph.Nummer}>
+      <div className="ds-stack ds-stack-8">
+        <Heading
+          tagName="h3"
+          text={`§ ${paragraph.Nummer} ${paragraph.Gesetz}`}
+          look="ds-subhead"
+          className="font-bold"
+        />
+        <p className="ds-subhead font-bold">{paragraph.Titel}</p>
+        <div className="ds-stack ds-stack-16 border-l-4 border-gray-400 pl-8">
+          {groupedAbsaetze.map((absatzGroup) =>
+            "id" in absatzGroup ? (
+              <Absatz key={absatzGroup.id} absatz={absatzGroup} />
+            ) : (
+              <DetailsSummary
+                key={absatzGroup[0].number}
+                title={getAbsatzGroupTitle(absatzGroup)}
+                bold={false}
+                content={
+                  <div className="ds-stack ds-stack-8">
+                    {absatzGroup.map((absatz) => (
+                      <BlocksRenderer
+                        key={absatz.id}
+                        content={prependNumberToAbsatz(absatz)}
+                        modifiers={{
+                          underline: PrincipleHighlight,
+                        }}
+                      />
+                    ))}
+                  </div>
+                }
+              />
+            ),
+          )}
+        </div>
+      </div>
+    </div>
+  );
+});
+
+const baseIDContext = createContext<string | null>(null);
+
+const Absatz = ({ absatz }: { absatz: AbsatzWithNumber }) => {
+  // This ID is used to label the reference in the highlight with the general explanation header
+  // and also serves as a basis for the link between the highlight and the specific explanation
+  const baseID = `erklaerung-${absatz.id}`;
```

Comment:
Maybe a more explicit name for the base ID could help since it is fetched from context at different places where the comment and context is not directly visible.